src/nvim/message.c

```diff
@@ -3145,6 +3167,71 @@ void msg_check(void)
   }
 }
 
+bool msg_enable_ext(void)
+{
+  return ui_has(kUIMessages) || msg_enable_msgfunc();
+}
+
+bool msg_enable_msgfunc(void)
+{
+  // msgfunc is disabled in command line mode
+  // Because default echo is used for command line redraw
+  return *p_msgfunc != NUL && !msg_check_loop
+    && !redir_off && !need_wait_return
+    && State != CMDLINE && State != CONFIRM;
+}
+
+static void msg_call_msgfunc_event(void **argv)
+{
+  char_u *msgfunc = argv[0];
+  typval_T *args = argv[1];
+
+  // Prevent infinite loop
+  msg_check_loop = true;
+  call_func_retnr(msgfunc, 4, args);
+  msg_check_loop = false;
+
+  xfree(msgfunc);
+
+  XFREE_CLEAR(args[1].vval.v_string);
+  xfree(args);
+
+  // ui_call_msg_xxx() calls ui_event(), which frees the allocated chunk
+  // The allocated chunk should also need freeing in or after
+  // msg_call_msgfunc().
+  api_free_array(msg_ext_chunks);
```

Comment:
According to https://github.com/neovim/neovim/pull/16480#discussion_r762515967 it's better to free this eagerly, right after `msg_call_msgfunc("msg_history_show", &entries);`, so this code can be removed.

Reply:
OK.